Malformed C code produced

Malformed C code is produced by the following snippet, possibly since the generic container BreadthOrder[V] is instantiated with the generic container LevelNode[V].

### Example

https://play.nim-lang.org/#ix=23TG

### Current Output
```
/usercode/nimcache/@min.nim.c: In function 'depthOf__fZLxCgVBTPuD5SUxZVsgmw':
/usercode/nimcache/@min.nim.c:800:83: error: incompatible type for argument 4 of 'expand__9aq8s76XX9a9aOsC1VbesSIVg'
    expand__9aq8s76XX9a9aOsC1VbesSIVg(order, tree, (*colonenv_).depthNode1.Field0, T13_);
                                                                                   ^~~~
In file included from /usercode/nimcache/@min.nim.c:9:0:
/usercode/nimcache/@min.nim.c:576:31: note: expected 'tyProc__vk6TkqumpYpdY6LdDcjUOw {aka struct <anonymous>}' but argument is of type 'tyProc__7qzvFW21g9cfwMABzEVWIJg {aka struct <anonymous>}'
 N_LIB_PRIVATE N_NIMCALL(void, expand__9aq8s76XX9a9aOsC1VbesSIVg)(tyObject_BreadthOrdercolonObjectType___PeOw2jeGEroF9bPDVOvXSNQ** order, tyObject_CappedStringTreecolonObjectType___XErj9cD0lQkSZ9a1q0WeoGcg* tree, NimStringDesc* node, tyProc__vk6TkqumpYpdY6LdDcjUOw transform) { nimfr_("expand", "/usercode/in.nim");
                               ^
/playground/nim/lib/nimbase.h:245:44: note: in definition of macro 'N_NIMCALL'
 #  define N_NIMCALL(rettype, name) rettype name /* no modifier */
```

### Expected Output
```
3
```

### Possible Solution

* Unknown

### Additional Information

* This resulted while I was trying to find a workaround to https://github.com/nim-lang/Nim/issues/12841 by generating code for different types of tree orderings in some experiments writing some abstract tree/graph algorithms.

```
$ nim -v
Nim Compiler Version 1.1.1 [Windows: amd64]
Compiled at 2019-12-04
```

Also occurs on 1.0.4 as seen in the play.nim-lang.org copy.

--- a/tests/concepts/tconcepts_issues.nim
+++ b/tests/concepts/tconcepts_issues.nim
@@ -525,3 +525,58 @@ block: # bug #21263
 
   print(10.Date) # ok
   print(DateDayFractionImpl(date: 1, fraction: 2))  # error
+
+import sets
+import deques
+
+type AnyTree[V] = concept t, type T
+  for v in t.leaves(V):
+    v is V
+
+type BreadthOrder[V] = ref object
+  frontier: Deque[V]
+  visited: HashSet[V]
+
+proc expand[V, T](order: ref BreadthOrder[T], tree: AnyTree[V], node: V, transform: (V) -> (T)) =
+  for leaf in tree.leaves(node):
+    if not order.visited.containsOrIncl(transform(leaf)):
+      order.frontier.addLast(transform(leaf))
+
+proc hasNext[V](order: ref BreadthOrder[V]): bool =
+  order.frontier.len > 0
+
+proc init[V](_: typedesc[BreadthOrder]): ref BreadthOrder[V] =
+  result.new()
+  result[] = BreadthOrder[V](frontier: initDeque[V](), visited: initHashSet[V]())
+
+proc popNext[V](order: ref BreadthOrder[V]): V =
+  order.frontier.popFirst()
+
+type LevelNode[V] = tuple
+  depth: uint
+  node: V
+
+proc depthOf*[V](orderType: typedesc[BreadthOrder], tree: AnyTree[V], root, goal: V): uint =
+  if root == goal:
+    return 0
+  var order = init[LevelNode[V]](orderType)
+  order.expand(tree, root, (leaf) => (1, leaf))
+  while order.hasNext():
+    let depthNode: LevelNode[V] = order.popNext()
+    if depthNode.node == goal:
+      return depthNode.depth
+    order.expand(tree, depthNode.node, (leaf) => (depthNode.depth + 1, leaf))
+
+type CappedStringTree = ref object
+  symbols: string
+  cap: Natural
+
+iterator leaves*(t: CappedStringTree, s: string): string =
+  if s.len < t.cap:
+    for c in t.symbols:
+      yield s & c
+
+block: # bug #12852
+  var tree = CappedStringTree(symbols: "^v><", cap: 5)
+
+  doAssert BreadthOrder.depthOf(tree, "", ">>>") == 3

